### PR TITLE
client: correctly unwrap getcwd dentries

### PR DESCRIPTION
**Pull request by batrick**
Mostly this is a fix for unwrapping dentry names for getcwd. I've also included a second fix for mkdir and recursive path walks needed for userspace encryption (see #61137).

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/70407](https://tracker.ceph.com/issues/70407).

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test windows

**Comment by batrick** [comment]
Previous code caused:

```
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:=================================== FAILURES ===================================
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:_______________________________ test_delete_cwd ________________________________
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:testdir = None
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:    def test_delete_cwd(testdir):
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:        assert_equal(b"/", cephfs.getcwd())
2025-03-12T10:49:57.566 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:        cephfs.mkdir(b"/temp-directory", 0o755)
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:        cephfs.chdir(b"/temp-directory")
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:        cephfs.rmdir(b"/temp-directory")
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:        # getcwd gives you something stale here: it remembers the path string
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:        # even when things are unlinked.  It's up to the caller to find out
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:        # whether it really still exists
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:>       assert_equal(b"/temp-directory", cephfs.getcwd())
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:../../../clone.client.0/src/test/pybind/test_cephfs.py:314:
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:a = b'/temp-directory', b = b'/'
2025-03-12T10:49:57.567 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:    def assert_equal(a, b):
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:>       assert a == b
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:E       AssertionError
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:../../../clone.client.0/src/test/pybind/assertions.py:2: AssertionError
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:---------------------------- Captured stdout setup -----------------------------
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:b'Purge /'
2025-03-12T10:49:57.568 INFO:tasks.workunit.client.0.smithi055.stdout:ret_buf=b''
```

/teuthology/pdonnell-2025-03-12_03:15:00-fs-wip-pdonnell-testing-20250312.000233-debug-distro-default-smithi/8182736/teuthology.log

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/70426](https://tracker.ceph.com/issues/70426).

**Comment by batrick** [comment]
jenkins test windows

File: src/test/libcephfs/test.cc [hunk 1]
```diff
@@ -562,6 +562,55 @@ TEST(LibCephFS, ManyNestedDirs) {
...
+  for (auto& component : mfp) {
+    struct ceph_dir_result *dirp;
+    ASSERT_EQ(ceph_opendir(cmount, ".", &dirp), 0);
+    struct dirent *dent = ceph_readdir(cmount, dirp);
+    ASSERT_TRUE(dent != NULL);
+    ASSERT_STREQ(dent->d_name, ".");
+    dent = ceph_readdir(cmount, dirp);
+    ASSERT_TRUE(dent != NULL);
+    ASSERT_STREQ(dent->d_name, "..");
+    if (component == "ManyNestedDirs"sv) {
```
Comment (vshankar) [inline src/test/libcephfs/test.cc:587]: Shouldn't this check be for `ManyNestedDirsCaseInsensitive` string?

File: src/client/Client.cc [hunk 2]
```diff
@@ -12333,7 +12329,7 @@ void Client::_getcwd(string& dir, const UserPerm& perms)
 
     // A cwd or ancester is unlinked
     if (in->dentries.empty()) {
-      return;
+      return -ENOENT;
```
Comment (vshankar) [inline src/client/Client.cc:12332]: Would this be the appropriate behaviour if a tree is deleted but the cwd is somewhere in that subtree? Maybe the result (of getcwd) if undefined, but its probably not -ENOENT.

File: src/libcephfs.cc [hunk 3]
```diff
@@ -322,7 +322,15 @@ struct ceph_mount_info
 
   int chdir(const char *to, const UserPerm& perms)
   {
-    return client->chdir(to, perms);
+    int rc = client->chdir(to, perms);
+    if (rc == 0) {
+      /* Current API requires "cwd" to be refreshed after every chdir so that
+       * getcwd on an unlinked cwd will still return the old path. Note:
+       * Client::getcwd now returns an error but leaves the "cwd" string
```
Comment (vshankar) [inline src/libcephfs.cc:329]: OK. So now I read this comment back-referring to getcwd() returning -ENOENT. While I'm fine with the refactor, why is it an optimization?

File: src/test/libcephfs/test.cc (hunk 1 above)
Comment (batrick) [reply src/test/libcephfs/test.cc:587]: Yes, thanks!

File: src/client/Client.cc (hunk 2 above)
Comment (batrick) [reply src/client/Client.cc:12332]: I checked on Linux and it will return -ENOENT:

```
$ mkdir foo
$ cd foo
~/foo$ rmdir ~/foo
~/foo$ strace -f -vv -- /bin/pwd
...
getcwd(0x628b23f84410, 4096)            = -1 ENOENT (No such file or directory)
```

File: src/libcephfs.cc (hunk 3 above)
Comment (batrick) [reply src/libcephfs.cc:329]: Well, the optimization was to stop doing a pointless `Client::getcwd` with every `::chdir` but apparently that was for libcephfs's benefit for a really subtle corner case. So I standby that I "optimized" the `Client` and pushed the corner-case handling into `libcephfs` where it belongs.

**Review by vshankar — APPROVED** [review]

File: src/test/libcephfs/test.cc [hunk 4]
```diff
@@ -521,40 +522,91 @@ TEST(LibCephFS, ManyNestedDirs) {
...
+    ASSERT_EQ(ceph_opendir(cmount, ".", &dirp), 0);
     struct dirent *dent = ceph_readdir(cmount, dirp);
     ASSERT_TRUE(dent != NULL);
     ASSERT_STREQ(dent->d_name, ".");
     dent = ceph_readdir(cmount, dirp);
     ASSERT_TRUE(dent != NULL);
     ASSERT_STREQ(dent->d_name, "..");
+    if (component == "ManyNestedDirsCaseInsensitive"sv) {
+      ASSERT_EQ(0, ceph_chdir(cmount, component.c_str()));
+      continue;
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/test/libcephfs/test.cc:589]: true, thanks
